translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,435 @@
+		sb.WriteString(entry.Value)
+	}
+	return data.TraceState(sb.String())
+}
+
+func ocAttrsToInternal(ocAttrs *octrace.Span_Attributes) data.Attributes {
+	if ocAttrs == nil {
+		return data.Attributes{}
+	}
+
+	attrCount := len(ocAttrs.AttributeMap)
+	attrMap := make(data.AttributesMap, attrCount)
+	if attrCount > 0 {
+		values := data.NewAttributeValueSlice(attrCount)
+		i := 0
+		for key, ocAttr := range ocAttrs.AttributeMap {
+			switch attribValue := ocAttr.Value.(type) {
+			case *octrace.AttributeValue_StringValue:
+				values[i].SetString(attribValue.StringValue.GetValue())
+
+			case *octrace.AttributeValue_IntValue:
+				values[i].SetInt(attribValue.IntValue)
+
+			case *octrace.AttributeValue_BoolValue:
+				values[i].SetBool(attribValue.BoolValue)
+
+			case *octrace.AttributeValue_DoubleValue:
+				values[i].SetDouble(attribValue.DoubleValue)
+
+			default:
+				str := "<Unknown OpenCensus attribute value type>"
+				values[i].SetString(str)
+			}
+			attrMap[key] = values[i]
+			i++
+		}
+	}
+	droppedCount := uint32(ocAttrs.DroppedAttributesCount)
+	return data.NewAttributes(attrMap, droppedCount)
+}
+
+func ocSpanKindToInternal(ocKind octrace.Span_SpanKind, ocAttrs *octrace.Span_Attributes) data.SpanKind {
+	switch ocKind {
+	case octrace.Span_SERVER:
+		return data.SpanKindSERVER
+
+	case octrace.Span_CLIENT:
+		return data.SpanKindCLIENT
+
+	case octrace.Span_SPAN_KIND_UNSPECIFIED:
+		// Span kind field is unspecified, check if TagSpanKind attribute is set.
+		// This can happen if span kind had no equivalent in OC, so we could represent it in
+		// the SpanKind. In that case the span kind may be a special attribute TagSpanKind.
+		if ocAttrs != nil {
+			kindAttr := ocAttrs.AttributeMap[tracetranslator.TagSpanKind]
+			if kindAttr != nil {
+				strVal, ok := kindAttr.Value.(*octrace.AttributeValue_StringValue)
+				if ok && strVal != nil {
+					var otlpKind data.SpanKind
+					switch tracetranslator.OpenTracingSpanKind(strVal.StringValue.GetValue()) {
+					case tracetranslator.OpenTracingSpanKindConsumer:
+						otlpKind = data.SpanKindCONSUMER
+					case tracetranslator.OpenTracingSpanKindProducer:
+						otlpKind = data.SpanKindPRODUCER
+					default:
+						return data.SpanKindUNSPECIFIED
+					}
+					delete(ocAttrs.AttributeMap, tracetranslator.TagSpanKind)
+					return otlpKind
+				}
+			}
+		}
+		return data.SpanKindUNSPECIFIED
+
+	default:
+		return data.SpanKindUNSPECIFIED
+	}
+}
+
+func ocEventsToInternal(ocEvents *octrace.Span_TimeEvents) (events []*data.SpanEvent, droppedCount uint32) {
+	if ocEvents == nil {
+		return
+	}
+
+	droppedCount = uint32(ocEvents.DroppedMessageEventsCount + ocEvents.DroppedAnnotationsCount)
+
+	eventCount := len(ocEvents.TimeEvent)
+	if eventCount == 0 {
+		return
+	}
+
+	events = data.NewSpanEventSlice(eventCount)
+
+	i := 0
+	for _, ocEvent := range ocEvents.TimeEvent {
+		if ocEvent == nil {
+			// Skip nil source events.
+			continue
+		}
+
+		event := events[i]
+		i++
```

Comment:
event := events[i++]?

Reply:
Not in Go, that's C/C++ :-)
In Go postfix increment is a statement, not an epression.